x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/WeightedAvg.java

```diff
@@ -0,0 +1,102 @@
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.multivalue.MvAvg;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.arithmetic.Div;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.arithmetic.Mul;
+
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+
+public class WeightedAvg extends AggregateFunction implements SurrogateExpression {
+
+    private final Expression weight;
+
+    @FunctionInfo(returnType = "double", description = "The weighted average of a numeric field.", isAggregation = true)
+    public WeightedAvg(
+        Source source,
+        @Param(name = "number", type = { "double", "integer", "long" }, description = "A numeric value.") Expression field,
+        @Param(name = "weight", type = { "double", "integer", "long" }, description = "A numeric weight.") Expression weight
+    ) {
+        super(source, field, List.of(weight));
+        this.weight = weight;
+    }
+
+    @Override
+    protected Expression.TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isType(
+            field(),
+            dt -> dt.isNumeric() && dt != DataType.UNSIGNED_LONG,
+            sourceText(),
+            FIRST,
+            "numeric except unsigned_long or counter types"
+        );
+
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return isType(
+            weight,
+            dt -> dt.isNumeric() && dt != DataType.UNSIGNED_LONG,
+            sourceText(),
+            SECOND,
+            "numeric except unsigned_long or counter types"
+        );
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.DOUBLE;
+    }
+
+    @Override
+    protected NodeInfo<WeightedAvg> info() {
+        return NodeInfo.create(this, WeightedAvg::new, field(), weight);
+    }
+
+    @Override
+    public WeightedAvg replaceChildren(List<Expression> newChildren) {
+        return new WeightedAvg(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    public Expression surrogate() {
+        var s = source();
+        var field = field();
+        var weight = weight();
+
+        if (field.foldable()) {
+            return new MvAvg(s, field);
+        }
+        if (weight.foldable()) {
+            return new Div(s, new Sum(s, field), new Count(s, field), dataType());
+        } else {
+            return new Div(s, new Sum(s, new Mul(s, field, weight)), new Sum(s, weight), dataType());
```

Comment:
Thought (absolutely not in scope here): When aggregating, it's normally okay when the field is multivalued and I _think_ the result of aggs is the same as if the field was mv_expand'ed beforehand. I wonder if weighted averages might be more useful if the substitution behaved the same way.
```
| MV_AVG(field, weights)
->
| SUM( MV_SUM(field) * weight) / SUM(MV_COUNT(field) * weight)

or even ->
| SUM( MV_SUM(field) * MV_SUM(weight) / SUM( MV_COUNT(field) * MV_SUM(weight))
```

Reply:
I was also thinking of how to deal with multi-valued field in the aggregation function, or even have a mv_weighted_agg? It might be confusing when combining them together, perhaps keep it simple for now and expand it when we see more use cases.